src/contracts/fungible/cache/sql.rs

```diff
@@ -10,3 +10,683 @@
 // You should have received a copy of the MIT License
 // along with this software.
 // If not, see <https://opensource.org/licenses/MIT>.
+
+use diesel::prelude::*;
+use std::collections::HashMap;
+use std::{fmt, fs, fs::File};
+
+use crate::contracts::fungible::cache::schema as cache_schema;
+
+use lnpbp::bitcoin;
+use lnpbp::bitcoin_hashes::hex::FromHex;
+use lnpbp::rgb::ContractId;
+
+use cache_schema::sql_allocation_utxo::dsl::sql_allocation_utxo as sql_allocation_utxo_table;
+use cache_schema::sql_allocations::dsl::sql_allocations as sql_allocation_table;
+use cache_schema::sql_assets::dsl::sql_assets as sql_asset_table;
+use cache_schema::sql_inflation::dsl::sql_inflation as sql_inflation_table;
+use cache_schema::sql_issues::dsl::sql_issues as sql_issue_table;
+
+use crate::contracts::fungible::data::Asset;
+
+use crate::error::{BootstrapError, ServiceErrorDomain};
+use std::path::PathBuf;
+
+use super::cache::{Cache, CacheError};
+
+use super::models::*;
+
+#[derive(Debug, Display, Error, From)]
+#[display(Debug)]
+pub enum SqlCacheError {
+    #[from]
+    Io(std::io::Error),
+
+    #[from]
+    Sqlite(diesel::result::Error),
+
+    #[from(bitcoin::hashes::hex::Error)]
+    HexDecoding,
+
+    #[from]
+    Generic(String),
+
+    #[from]
+    BlindKey(lnpbp::secp256k1zkp::Error),
+
+    #[from]
+    WrongChainData(lnpbp::bp::chain::ParseError),
+
+    #[from(std::option::NoneError)]
+    NotFound,
+}
+
+impl From<SqlCacheError> for ServiceErrorDomain {
+    fn from(_: SqlCacheError) -> Self {
+        ServiceErrorDomain::Cache
+    }
+}
+
+impl From<SqlCacheError> for BootstrapError {
+    fn from(_: SqlCacheError) -> Self {
+        BootstrapError::CacheError
+    }
+}
```

Comment:
This can be removed and replaced with two `#[from]` directives in `ServiceErrorDomain` declaration

Reply:
Good point. Should it also work for `FileCacheError`? And how should i write conversion derivation from multiple error types?

Reply:
Conversion for multiple types works with multiple and separate `#[from(...)]` attributes.

Reply:
And should the same thing could be done for `FileCacheError`?

Reply:
Yes

Reply:
Done. Added the same for filecache error too.